Create mongo-high-connections.md (#53)
percona-platform/portal-doc: docs/advisors/checks/mongo-high-connections.md

# MongoDB High Connections

## Description

> This check returns warnings if there is an increase in the number of connections.

Reviewer: ```suggestion
This check warns if the number of connections is increasing.
```

Reviewer 2: I think, the above statement should be -
"This check warns if the number of connections are increased."

Reviewer: Hmm, the agreement is with "the number" here.

## Resolution

> High connection count in MongoDB can be caused by various factors, including application design, resource utilization, and server configuration. Here are some common reasons and potential solutions for high connection counts in MongoDB:

Reviewer: ```suggestion
High connection count in MongoDB can be caused by various factors, including application design, resource utilization, and server configuration. 
Here are some common reasons and potential solutions for high connection counts in MongoDB:
```

-  **Connection pooling issues**

>    Connection pooling is a technique used to manage database connections efficiently. If the connection pool is not configured correctly, it can lead to high connection counts in MongoDB. You can check the maximum size of the connection pool and adjust it as needed. Additionally, you should check the connection timeout values to ensure that connections are being reused effectively.

Reviewer: ```suggestion
   Connection pooling is a technique used to manage database connections efficiently. Not configuring the connection pool correctly can lead to high connection counts in MongoDB. 
You can check the maximum size of the connection pool and adjust it as needed. Additionally, you should check the connection timeout values to ensure that connections are being reused effectively.
```

   
-  **Connection leaks**

>   Connection leaks occur when a connection is not closed properly, leading to the creation of additional connections over time. To prevent connection leaks, you can ensure that your application closes connections after they are no longer needed.

Reviewer: ```suggestion
  Connection leaks occur when a connection is not closed correctly. Over time, this can lead to additional connections being created.  To prevent connection leaks, make sure that your application closes connections after they are no longer needed.
```

> -  **Batch Jobs**

Reviewer: ```suggestion
-  **Batch jobs**
```

>    There might be regular batch jobs running which increase the number of connections and also the resource utilization. Additionally, you can verify the kind of operations used by running the mongostats command.

Reviewer: ```suggestion
   Regular batch jobs increase the number of connections and impact resource utilization. Additionally, you can verify the kind of operations used by running the **mongostats** command.
```

-  **Poorly optimized queries**

>    Queries that perform full table scans or involve a large number of documents can consume a significant amount of resources, resulting in high connection counts since the queues will be increased. To reduce connection counts caused by queries, it's essential to optimize your queries. You can use the MongoDB Explain() method to understand query performance and identify potential areas for optimization.

Reviewer: ```suggestion
   Queries that perform full table scans or involve a large number of documents can consume a significant amount of resources. Since the queues will be increased, this can result in high connection counts. 
To reduce connection counts caused by queries, it's essential to optimize your queries. You can use the **MongoDB Explain()** method to understand query performance and identify potential areas for optimization.
```

-  **Application design issues**

>   If the application has been designed to create too many connections, it can lead to high connection counts in MongoDB. You can consider implementing connection sharing, where connections are reused across application instances, to minimize connection overhead.

Reviewer: ```suggestion
If the application has been designed to create too many connections, it can lead to high connection counts in MongoDB. To minimize connection overhead, consider implementing connection sharing, where connections are reused across application instances.
```

-  **Server resource constraints**

>    Insufficient server resources such as CPU, memory, or disk I/O can lead to high connection counts in MongoDB. You can check the system resource utilization and increase resources as needed. You may also want to consider adding more servers to the replica set to distribute the load.

Reviewer: ```suggestion
Insufficient server resources such as CPU, memory, or disk I/O can lead to high connection counts in MongoDB. Check the system resource utilization and increase resources as needed. You may also want to consider adding more servers to the replica set to distribute the load.
```
